pkg/kubectl/cmd/util/factory_object_mapping.go

```diff
@@ -427,3 +438,41 @@ func (f *ring1Factory) SwaggerSchema(gvk schema.GroupVersionKind) (*swagger.ApiD
 	}
 	return discovery.SwaggerSchema(version)
 }
+
+// OpenAPISchema returns metadata and structural information about Kubernetes object definitions.
+// Will try to cache the data to a local file.  Cache is written and read from a
+// file created with ioutil.TempFile and obeys the expiration semantics of that file.
+// The cache location is a function of the client and server versions so that the open API
+// schema will be cached separately for different client / server combinations.
+// Note, the cache will not be invalidated if the server changes its open API schema without
+// changing the server version.
+func (f *ring1Factory) OpenAPISchema(cacheDir string) (*openapi.Resources, error) {
+	discovery, err := f.clientAccessFactory.DiscoveryClient()
+	if err != nil {
+		return nil, err
+	}
+
+	// Lazily initialize the OpenAPIGetter once
+	f.openAPIGetter.Do(func() {
+		// Get the server version for caching the openapi spec
+		versionString := ""
+		version, err := discovery.ServerVersion()
+		if version != nil && err != nil {
+			// Cache the result under the server version
+			versionString = version.String()
+		}
+
+		// Get the cache directory for caching the openapi spec
+		cacheDir, err = substituteUserHome(cacheDir)
+		if err != nil {
+			// Don't cache the result if we couldn't subsistute the home directory
+			cacheDir = ""
+		}
+
+		// Create the caching OpenAPIGetter
+		f.openAPIGetter.Getter = openapi.NewOpenAPIGetter(cacheDir, versionString, discovery)
```

Comment:
Does the discovery client expose a method to that will give the client stable a hash / key for for exposed API?